src/senzing_pipeline.py

```diff
@@ -145,29 +188,236 @@ def generate_summaries(
+                records[record_uid] = match_key
+
+                if entity_name == "" and rec["ENTITY_DESC"] != "":
+                    entity_name = rec["ENTITY_DESC"]
+
+            if entity_name == "":
+                entity_name = str(entity_uid)
+
+            entities[entity_uid] = Entity(
+                entity_uid=entity_uid,
+                name=entity_name,
+                records=records,
+                num_recs=len(records),
+                related={r["ENTITY_ID"]: r for r in entity_dat["RELATED_ENTITIES"]},
+            )
+
+    for entity in entities.values():
+        if entity.num_recs > 0:
+            entity.has_ref = True
+
+        for rel_ent_id in entity.related:
+            entities[rel_ent_id].has_ref = True
+
+    return entities
+
+
+def filter_senzing() -> set[str]:
+    logger.info("Loading suspicions")
+    with open("data/icij-example/suspicious.txt") as file:
+        names = [line.rstrip() for line in file]
+
+    logger.info("Loading Senzing results")
+    entities = extract_senzing_results("data/ICIJ-entity-report-2024-06-21_12-04-57-std.json")
+
+    logger.info("Filter Senzing results for suspicions only")
+    df = pd.DataFrame.from_records(
+        [
+            (
+                name,
+                list(
+                    # here we filter for exact match, which is high precision + low recall
+                    # we might miss companies with names that differ slightly
+                    # we hope to catch them with the "friends of friends" filtering later
+                    filter(lambda ent: ent.name == name, [entity for k, entity in entities.items()])
+                ),
+            )
+            for name in names
+        ],
+        columns=["suspicion", "matches"],
+    )
+
+    # we filter Senzing results for: direct matches, related entities to direct matches and friends of friends
+    rank_0 = df.matches.explode().dropna().apply(lambda d: d.entity_uid).unique()
+    rank_1 = df.matches.explode().dropna().apply(lambda d: d.related.keys()).explode().unique()
+    rank_2 = set(
+        ent_id
+        for seed_id in set(rank_0) | set(rank_1)
+        for ent_id in entities[seed_id].related.keys()
+    )
+    return set(str(ent_id) for ent_id in set(rank_0) | set(rank_1) | set(rank_2))
 
 
 def main():
     """Entrypoint to the Senzing data pipeline."""
     countries = load_countries()
-    entities = load_entities()
-    summaries = generate_summaries(entities, countries)
-    write_summaries(summaries)
+    raw_entities = load_entities()
+    raw_aliases = load_aliases()
+
+    entity_ids = filter_senzing()
+    filtered_entities = {k: v for k, v in raw_entities.items() if str(k) in entity_ids}
+    filtered_aliases = [alias for alias in raw_aliases if str(alias["entity"]) in entity_ids]
+
+    entities = generate_entities(filtered_entities, countries)
+    write_entities(entities)
+
+    aliases = generate_aliases(filtered_aliases)
+    write_aliases(aliases)
```

Comment:
this is the main Data Engineering pipeline. In a future PR, I might clean it up, for the purpose of illustrating it and explaining it in the article